C error on compare aliased arrays with initialisation: `Arr{} == Arr{}`

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->

**V doctor:**

```
V full version: V 0.4.8 5ec9bb5.610558d
OS: linux, Linux version 6.6.8-arch1-1 (linux@archlinux) (gcc (GCC) 13.2.1 20230801, GNU ld (GNU Binutils) 2.41.0) #1 SMP PREEMPT_DYNAMIC Thu, 21 Dec 2023 19:01:01 +0000
Processor: 16 cpus, 64bit, little endian, 12th Gen Intel(R) Core(TM) i5-1240P

getwd: /home/ge/Code/v/netaddr/issues
vexe: /home/ge/.vlang/v
vexe mtime: 2024-12-13 10:58:51

vroot: OK, value: /home/ge/.vlang
VMODULES: OK, value: /home/ge/.vmodules
VTMP: OK, value: /tmp/v_1000

Git version: git version 2.47.1
Git vroot status: weekly.2024.50-22-g610558d4
.git/config present: true

CC version: cc (GCC) 14.2.1 20240910
emcc version: N/A
thirdparty/tcc status: thirdparty-linux-amd64 0134e9b9
```

**What did you do?**
`./v -g -o vdbg cmd/v && ./vdbg initialized_aliased_arrays_comparison.vv`

```v
type Arr = [4]u8

a := Arr{}
b := Arr{}

// works fine
assert a == b

// C error on each below

assert Arr{} == Arr{}

assert Arr{} == [4]u8{}

assert Arr{} == [u8(0), 0, 0, 0]!
```

**What did you expect to see?**

Assertions passed, no C errors

**What did you see instead?**

```
================== C compilation error (from tcc): ==============
cc: /tmp/v_1000/initialized_aliased_arrays_comparison.01JEZXWR0N6F9QFTPX5ED09YM8.tmp.c:5883: error: expression expected before '{'
=================================================================
(You can pass `-cg`, or `-show-c-output` as well, to print all the C error messages).
builder error: 
==================
C error found. It should never happen, when compiling pure V code.
This is a V compiler bug, please report it using `v bug file.v`,
or goto https://github.com/vlang/v/issues/new/choose .
You can also use #help on Discord: https://discord.gg/vlang .
```

> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

<sub><a href="https://huly.app/guest/vlang-66f40c4d-a476b54c67-771fdd?token=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJsaW5rSWQiOiI2NzVjMjA4YjFlMDYzMmRkMDhlMzc5ZDEiLCJndWVzdCI6InRydWUiLCJlbWFpbCI6IiNndWVzdEBoYy5lbmdpbmVlcmluZyIsIndvcmtzcGFjZSI6InctYWxleGFuZGVyLXZsYW5nLTY2ZjQwYzRkLWE0NzZiNTRjNjctNzcxZmRjIn0.aNG-5OUFPueEmg0ywsMrvxw4IMLQcpaDLx9bSHf9UhQ">Huly&reg;: <b>V_0.6-21586</b></a></sub>

--- a/vlib/v/ast/types.v
+++ b/vlib/v/ast/types.v
@@ -1006,6 +1006,16 @@ pub fn (t &Struct) is_unresolved_generic() bool {
 	return t.generic_types.len > 0 && t.concrete_types.len == 0
 }
 
+pub fn (t &TypeSymbol) is_primitive_fixed_array() bool {
+	if t.info is ArrayFixed {
+		return global_table.final_sym(t.info.elem_type).is_primitive()
+	} else if t.info is Alias {
+		return global_table.final_sym(t.info.parent_type).is_primitive_fixed_array()
+	} else {
+		return false
+	}
+}
+
 pub fn (t &TypeSymbol) is_array_fixed() bool {
 	if t.info is ArrayFixed {
 		return true

--- a/vlib/v/gen/c/auto_eq_methods.v
+++ b/vlib/v/gen/c/auto_eq_methods.v
@@ -437,6 +437,11 @@ fn (mut g Gen) gen_fixed_array_equality_fn(left_type ast.Type) string {
 
 	mut fn_builder := strings.new_builder(512)
 	fn_builder.writeln('inline bool ${ptr_styp}_arr_eq(${arg_styp} a, ${arg_styp} b) {')
+	if left_typ.sym.is_primitive_fixed_array() {
+		fn_builder.writeln('\tif (!memcmp(${left}, ${right}, sizeof(${arg_styp}))) {')
+		fn_builder.writeln('\t\treturn true;')
+		fn_builder.writeln('\t}')
+	}
 	fn_builder.writeln('\tfor (int i = 0; i < ${size}; ++i) {')
 	// compare every pair of elements of the two fixed arrays
 	if elem.sym.kind == .string {

--- a/vlib/v/gen/c/infix.v
+++ b/vlib/v/gen/c/infix.v
@@ -216,13 +216,13 @@ fn (mut g Gen) infix_expr_eq_op(node ast.InfixExpr) {
 					if left.typ.is_ptr() {
 						g.write('*'.repeat(left.typ.nr_muls()))
 					}
-					if node.left is ast.StructInit {
+					if !g.is_cc_msvc && node.left is ast.StructInit {
 						s := g.styp(left.unaliased)
 						g.write('(${s})')
 					}
 					g.expr(node.left)
 					g.write(', ')
-					if node.right is ast.StructInit {
+					if !g.is_cc_msvc && node.right is ast.StructInit {
 						s := g.styp(right.unaliased)
 						g.write('(${s})')
 					}
@@ -274,8 +274,7 @@ fn (mut g Gen) infix_expr_eq_op(node ast.InfixExpr) {
 						s := g.styp(left.unaliased)
 						g.write('(${s})')
 					}
-				}
-				if node.left is ast.StructInit {
+				} else if !g.is_cc_msvc && node.left is ast.StructInit {
 					s := g.styp(left.unaliased)
 					g.write('(${s})')
 				}
@@ -286,8 +285,7 @@ fn (mut g Gen) infix_expr_eq_op(node ast.InfixExpr) {
 						s := g.styp(right.unaliased)
 						g.write('(${s})')
 					}
-				}
-				if node.right is ast.StructInit {
+				} else if !g.is_cc_msvc && node.right is ast.StructInit {
 					s := g.styp(right.unaliased)
 					g.write('(${s})')
 				}

--- a/vlib/v/gen/c/infix.v
+++ b/vlib/v/gen/c/infix.v
@@ -216,13 +216,14 @@ fn (mut g Gen) infix_expr_eq_op(node ast.InfixExpr) {
 					if left.typ.is_ptr() {
 						g.write('*'.repeat(left.typ.nr_muls()))
 					}
-					if !g.is_cc_msvc && node.left is ast.StructInit {
+					if node.left is ast.StructInit && left.unaliased_sym.is_primitive_fixed_array() {
 						s := g.styp(left.unaliased)
 						g.write('(${s})')
 					}
 					g.expr(node.left)
 					g.write(', ')
-					if !g.is_cc_msvc && node.right is ast.StructInit {
+					if node.right is ast.StructInit
+						&& right.unaliased_sym.is_primitive_fixed_array() {
 						s := g.styp(right.unaliased)
 						g.write('(${s})')
 					}
@@ -274,7 +275,8 @@ fn (mut g Gen) infix_expr_eq_op(node ast.InfixExpr) {
 						s := g.styp(left.unaliased)
 						g.write('(${s})')
 					}
-				} else if !g.is_cc_msvc && node.left is ast.StructInit {
+				} else if node.left is ast.StructInit
+					&& left.unaliased_sym.is_primitive_fixed_array() {
 					s := g.styp(left.unaliased)
 					g.write('(${s})')
 				}
@@ -285,7 +287,8 @@ fn (mut g Gen) infix_expr_eq_op(node ast.InfixExpr) {
 						s := g.styp(right.unaliased)
 						g.write('(${s})')
 					}
-				} else if !g.is_cc_msvc && node.right is ast.StructInit {
+				} else if node.right is ast.StructInit
+					&& right.unaliased_sym.is_primitive_fixed_array() {
 					s := g.styp(right.unaliased)
 					g.write('(${s})')
 				}

--- a/vlib/v/gen/c/infix.v
+++ b/vlib/v/gen/c/infix.v
@@ -216,8 +216,16 @@ fn (mut g Gen) infix_expr_eq_op(node ast.InfixExpr) {
 					if left.typ.is_ptr() {
 						g.write('*'.repeat(left.typ.nr_muls()))
 					}
+					if node.left is ast.StructInit {
+						s := g.styp(left.unaliased)
+						g.write('(${s})')
+					}
 					g.expr(node.left)
 					g.write(', ')
+					if node.right is ast.StructInit {
+						s := g.styp(right.unaliased)
+						g.write('(${s})')
+					}
 					if right.typ.is_ptr() {
 						g.write('*'.repeat(right.typ.nr_muls()))
 					}
@@ -267,6 +275,10 @@ fn (mut g Gen) infix_expr_eq_op(node ast.InfixExpr) {
 						g.write('(${s})')
 					}
 				}
+				if node.left is ast.StructInit {
+					s := g.styp(left.unaliased)
+					g.write('(${s})')
+				}
 				g.expr(node.left)
 				g.write(', ')
 				if node.right is ast.ArrayInit {
@@ -274,6 +286,10 @@ fn (mut g Gen) infix_expr_eq_op(node ast.InfixExpr) {
 						s := g.styp(right.unaliased)
 						g.write('(${s})')
 					}
+				}
+				if node.right is ast.StructInit {
+					s := g.styp(right.unaliased)
+					g.write('(${s})')
 				}
 				g.expr(node.right)
 				g.write(')')

--- a/vlib/v/gen/c/str.v
+++ b/vlib/v/gen/c/str.v
@@ -234,7 +234,7 @@ fn (mut g Gen) gen_expr_to_string(expr ast.Expr, etype ast.Type) {
 			if temp_var_needed {
 				g.write(tmp_var)
 			} else {
-				if expr is ast.StructInit && sym.kind in [.array_fixed, .alias] {
+				if !g.is_cc_msvc && expr is ast.StructInit && sym.kind in [.array_fixed, .alias] {
 					s := g.styp(expr.typ)
 					g.write('(${s})')
 				}

--- a/vlib/v/gen/c/str.v
+++ b/vlib/v/gen/c/str.v
@@ -234,7 +234,7 @@ fn (mut g Gen) gen_expr_to_string(expr ast.Expr, etype ast.Type) {
 			if temp_var_needed {
 				g.write(tmp_var)
 			} else {
-				if !g.is_cc_msvc && expr is ast.StructInit && sym.kind in [.array_fixed, .alias] {
+				if expr is ast.StructInit && g.table.final_sym(expr.typ).is_primitive_fixed_array() {
 					s := g.styp(expr.typ)
 					g.write('(${s})')
 				}

--- a/vlib/v/gen/c/str.v
+++ b/vlib/v/gen/c/str.v
@@ -234,6 +234,10 @@ fn (mut g Gen) gen_expr_to_string(expr ast.Expr, etype ast.Type) {
 			if temp_var_needed {
 				g.write(tmp_var)
 			} else {
+				if expr is ast.StructInit && sym.kind in [.array_fixed, .alias] {
+					s := g.styp(expr.typ)
+					g.write('(${s})')
+				}
 				g.expr_with_cast(expr, typ, typ)
 			}
 		} else if typ.has_flag(.option) {

--- a/vlib/v/tests/assert_alias_array_test.v
+++ b/vlib/v/tests/assert_alias_array_test.v
@@ -1,0 +1,15 @@
+type Arr = [4]u8
+type Arr2 = []u8
+
+fn test_main() {
+	a := Arr{}
+	b := Arr{}
+
+	assert a == b
+	assert Arr{} == Arr{}
+	assert Arr{} == [4]u8{}
+	assert Arr{} == [u8(0), 0, 0, 0]!
+
+	assert Arr2{} == Arr2{}
+	assert Arr2{} == []u8{}
+}

--- a/vlib/v/tests/assert_alias_array_test.v
+++ b/vlib/v/tests/assert_alias_array_test.v
@@ -1,5 +1,6 @@
 type Arr = [4]u8
 type Arr2 = []u8
+type ArrStr = [4]string
 
 fn test_main() {
 	a := Arr{}
@@ -12,4 +13,6 @@ fn test_main() {
 
 	assert Arr2{} == Arr2{}
 	assert Arr2{} == []u8{}
+
+	assert ArrStr{} == ArrStr{}
 }

